### DISCUSSION TITLE
DAI 2024-2025 - Java IOs benchmarking - Maxime Lestiboudois

### DISCUSSION THREAD

**Discussion by MaximeLesti** [heig-vd-dai-course/heig-vd-dai-course#367] [Show and tell]
https://github.com/MaximeLesti/heig-vd-dai-course-java-io

### **Lecture (READ)**

| Taille         | text      | buffered_text | binary     | buffered_binary |
|----------------|-----------|---------------|------------|-----------------|
| **1B**         | 166 ms    | 152 ms        | 160 ms     | 152 ms          |
| **1KiB**       | 177 ms    | 154 ms        | 144 ms     | 146 ms          |
| **1MiB**       | 234 ms    | 182 ms        | 713 ms     | 158 ms          |
| **5MiB**       | 440 ms    | 220 ms        | 2436 ms    | 220 ms          |

### **Écriture (WRITE)**

| Taille         | text      | buffered_text | binary     | buffered_binary |
|----------------|-----------|---------------|------------|-----------------|
| **1B**         | 139 ms    | 143 ms        | 137 ms     | 145 ms          |
| **1KiB**       | 152 ms    | 161 ms        | 138 ms     | 145 ms          |
| **1MiB**       | 145 ms    | 160 ms        | 128 ms     | 131 ms          |
| **5MiB**       | 161 ms    | 158 ms        | 151 ms     | 148 ms          |

- Which type of stream is the most efficient for each use case?
In all cases, the most efficient is the buffered_binary stream, even if binary is very efficient for writing.

- Why is it more efficient than the other types of streams?
Buffered_binary

- What is the difference between binary data and text data?
We don't care about the format of binary data (it is only 0 and 1). It exists many formats for the text (UTF-8, ASCII, etc.). Binary is computer-friendly and text is human-friendly.
The principal difference is how the data is encoded in the file.

- What is a character encoding?
It is a set of mappings between the bytes in the computer and the characters in the character set (ASCII, UTF-8)

- Why is this benchmark methodology important?
It is important to create optimal and efficient software. 
it's an opportunity to learn what could be optimized by comparing two processes.

**Comment by ludelafo** [comment]
Thank you!

You can always find the solution in the solution repository if you ever need to compare your work with the official one here: https://github.com/heig-vd-dai-course/heig-vd-dai-course-solutions.

Do not hesitate to ask questions if something is not clear, either in this discussion or in the Q&A category of the GitHub Discussions here: https://github.com/orgs/heig-vd-dai-course/discussions/categories/q-a.
